pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -335,29 +324,24 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 		return reconcile.Result{}, err
 	}
 
-	pendingClaims, err := r.getAllPendingClusterClaims(clp, logger)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
-	logger.WithField("count", len(pendingClaims)).Debug("found pending claims for ClusterPool")
-
 	// reserveSize is the number of clusters that the pool currently has in reserve
-	reserveSize := len(installingCDs) + len(readyCDs) - len(pendingClaims)
+	reserveSize := len(cds.Installing()) + len(cds.Assignable()) - len(claims.Unassigned())
 
-	readyCDs, err = r.assignClustersToClaims(pendingClaims, readyCDs, logger)
-	if err != nil {
+	if err := assignClustersToClaims(r.Client, claims, cds, logger); err != nil {
+		logger.WithError(err).Error("error assigning clusters <=> claims")
 		return reconcile.Result{}, err
 	}
 
 	availableCurrent := math.MaxInt32
 	if clp.Spec.MaxConcurrent != nil {
-		availableCurrent = int(*clp.Spec.MaxConcurrent) - len(installingCDs) - numberOfDeletingCDs - numberOfDeletingClaimedCDs
+		availableCurrent = int(*clp.Spec.MaxConcurrent) - len(cds.Installing()) - len(cds.Deleting())
```

Comment:
what about the marked for deletion ? that seems to be have been removed from calculation..

Reply:
> https://github.com/openshift/hive/pull/1474/files#diff-bbf4d3cb24b8cac87951b84ca973747902c00d4a9d662648c8e760a37569581dR303-R307

> // Deleting returns the list of ClusterDeployments whose DeletionTimestamp is set. Not to be
> // confused with MarkedForDeletion.

that does not track with the comment message on the function?

Reply:
but seems like the previous code was looking for deletion timestamp set for pool managed cds, and already deleting claimed clusters. so the code itself seems good. 👍🏽

Reply:
Sorry, this

> As noted previously, Deleting() includes both deleting-per-deletionTimestamp and marked-for-deletion.

was totally wrong, as you pointed out. The method's docstring is accurate. What I meant to say, as you noted, was that `Deleting()` encompasses both *claimed and unclaimed* CDs with `deletionTimestamp` set.